src/audio_unit/looper.rs

```diff
@@ -88,24 +99,70 @@ impl Looper {
                 }
             }
             Playing { position, total } => {
+                self.process_samples_playing(position, total, input, output);
+            }
+            PlayingAwaitingOverdub { position, total } => {
+                self.process_samples_playing(position, total, input, output);
+            }
+            Overdubbing { position, total } => {
                 let next_position = position + output.len();
 
                 if next_position <= total {
                     output.copy_from_slice(&self.buffer[position..next_position]);
+
+                    for (sample_number, input_sample) in input.iter().enumerate() {
+                        self.buffer[position + sample_number] += input_sample;
+                    }
+
                     let position = if next_position == total {
                         0
                     } else {
                         next_position
                     };
 
-                    self.state = Playing { position, total }
+                    self.state = if next_position == total {
+                        println!("looper: done overdubbing");
+                        Playing { position, total }
+                    } else {
+                        Overdubbing { position, total }
+                    }
                 } else {
                     self.process_samples_wrap_around(position, total, input, output);
                 }
             }
         };
     }
 
+    fn process_samples_playing(
+        &mut self,
+        position: usize,
+        total: usize,
+        input: &[f32],
+        output: &mut [f32],
+    ) {
+        let next_position = position + output.len();
+
+        if next_position <= total {
+            output.copy_from_slice(&self.buffer[position..next_position]);
+            let position = if next_position == total {
+                0
+            } else {
+                next_position
+            };
+
+            self.state = match self.state {
+                PlayingAwaitingOverdub { .. } if next_position == total => {
+                    println!("looper: activating overdub");
+                    Overdubbing { position, total }
+                }
+                PlayingAwaitingOverdub { .. } => PlayingAwaitingOverdub { position, total },
+                _ => Playing { position, total },
+            }
```

Comment:
Same comment as above about considering whether this could be in a `impl State`.

Reply:
Reading this now after having refactored to `impl State`, this reads cleanly in a certain way but I'm not convinced that I like the "hiding" of the state enum values (this feels "slipperier" to me with `self.state` as a bit of a "black box" - "how do I know when to invoke one of these methods on it?")

Reply:
Yeah, I think that's fair. Since the `state` is related to the other state of this struct, maybe it doesn't really work to put a bunch of the logic into `State`.